main_test.go

```diff
@@ -713,3 +716,37 @@ func Test_runScorecardSettings(t *testing.T) {
 		})
 	}
 }
+
+func Test_signScorecardResult(t *testing.T) {
+	t.Parallel()
+	// Sign example scorecard results file.
+	scorecardResultsFile := "./testdata/scorecard-results-example.sarif"
+	err := signScorecardResult(scorecardResultsFile)
+	if err != nil {
+		t.Errorf("signScorecardResult() error, %v", err)
+		return
+	}
+
+	// Verify that the signature was created and uploaded to the Rekor tlog by looking up the payload.
+	ctx := context.Background()
+	rekorClient, err := rekor.NewClient("https://rekor.sigstore.dev")
+	if err != nil {
+		t.Errorf("signScorecardResult() error getting Rekor client, %v", err)
+		return
+	}
+	scorecardResultData, err := ioutil.ReadFile("./testdata/scorecard-results-example.sarif")
+	if err != nil {
+		t.Errorf("signScorecardResult() error reading scorecard result file, %v", err)
+		return
+	}
+	uuids, err := cosign.FindTLogEntriesByPayload(ctx, rekorClient, scorecardResultData)
+	if err != nil {
+		t.Errorf("signScorecardResult() error getting tlog entries, %v", err)
+		return
+	}
+
+	if len(uuids) == 0 {
```

Comment:
are you sure searching for the tlog entry via payload work? Since the test will run on each push/pull request, previous runs will have added to the log the same payload. So I suspect the test passes even if you comment out the call to `signScorecardResult()`. Can you verify? If that's the case, either match their signature (which is probabilistic if we use ECDSA). Another approach is to add randomness to the payload

Reply:
agreed, the test will always pass. 

> If that's the case, either match their signature (which is probabilistic if we use ECDSA).

We can return the signature from the `SignBlobCmd` and then in `signScorecardResult()`. You can iterate through the UUIDs and extract their signatures (see this similar [func](https://github.com/sigstore/cosign/blob/18d2ce0b458018951f7356db911467a427a8dffe/cmd/cosign/cli/verify/verify_blob.go#L345) that extracts the certificate from the hashedrekord payload, you want the signature payload), and check if one matches.

Reply:
I wonder whether it's better to randomize the payload for this use case. My worry is that the search is going to grow O(n^2) with n being he number of runs. This is not a problem for genuine releases because projects release a limited number of releases in their lifetime (although it'd still be nice for rekor to have a "commit has index" because it'd make things faster)

I first thought the commit hash index could also be a mitigation against DoS since anyone can add random entries with the same payload... but I guess anyone can add entries for the same hash too :-) It'd be nice to mitigate these sort of attack by having rekor aware of GitHub, and verify that the entry was generated from GitHub workflow before adding the entry to the log. I think we could add an additional flag in the cert, like "origin", but I suspect it's already possible to infer this given the cert? @asraa wdut?

Reply:
> . My worry is that the search is going to grow O(n) with n being the number of runs

You're right -- each test invocation is going to cause an additional UUID to search. Let's add a nonce to the payload.

> and verify that the entry was generated from GitHub workflow before adding the entry to the log.

Why should rekor prevent someone from uploading a verified entry? Or are you suggesting that if rekor sees a github workflow certificate in an entry, that it should verify that the upload was happening inside a workflow? I think I disagree -- rekor already verifies that the signature is valid on upload, so this is "probably" enough to mitigate a DoS -- you have to go through the process of signing before you can stick something in the log. It also de-dupes entries, so you can't spam rekor unless you went through the process of signing inside the workflow.

Reply:
I think adding some randomness to the payload is a good idea.

> Or are you suggesting that if rekor sees a github workflow certificate in an entry, that it should verify that the upload was happening inside a workflow?

Isn't it necessary to prove that the upload came from the scorecard workflow? What if someone tampered with the results and then used the same process to sign and upload them. Then, when verifying, we will lookup the payload and see a corresponding tlog entry.

Reply:
probably unrelated question: can rekor be queried by signature instead of payload? Note that there are also re-builders who may sign reproducible builds, so the payload may get duplicated: is that correct? What do you mean by `de-dupes entries`?

@asraa IIUC, rekor verifies the signature chains up to Fulcio root before allowing the entry to be added? (so it forces a user to perform at least the OIDC flow). How about non-OIDC flows?

who generates the public key, the client or Fulcio?